Firestore/Swift/Source/Codable/DocumentID.swift

```diff
@@ -65,57 +66,114 @@ internal protocol DocumentIDProtocol {
 }
 
-/// A value that is populated in Codable objects with the `DocumentReference`
-/// of the current document by the Firestore.Decoder when a document is read.
+/// A property wrapper type that marks a `DocumentReference?` or `String?` field to
+/// be populated with a document identifier when it is read.
 ///
-/// If the field name used for this type conflicts with a read document field,
-/// an error is thrown. For example, if a custom object has a field `firstName`
-/// annotated with `@DocumentID`, and there is a property from the document
-/// named `firstName` as well, an error is thrown when you try to read the
-/// document.
+/// Apply the `@DocumentID` annotation to a `DocumentReference?` or `String?`
+/// property in a `Codable` object to have it populated with the document
+/// identifier when it is read and decoded from Firestore.
 ///
-/// When writing a Codable object containing an `@DocumentID` annotated field,
-/// its value is ignored. This allows you to read a document from one path and
-/// write it into another without adjusting the value here.
+/// - Important: The name of the property annotated with `@DocumentID` must not
+///   match the name of any fields in the Firestore document being read or else
+///   an error will be thrown. For example, if the `Codable` object has a
+///   property named `firstName` annotated with `@DocumentID`, and the Firestore
+///   document contains a field named `firstName`, an error will be thrown when
+///   attempting to decode the document.
 ///
-/// NOTE: Trying to encode/decode this type using encoders/decoders other than
-/// Firestore.Encoder leads to an error.
+/// - Example Read:
+///   ````
+///   struct Player: Codable {
+///     @DocumentID var playerID: String?
+///     var health: Int64
+///   }
+///
+///   let p = try! await Firestore.firestore()
+///       .collection("players")
+///       .document("player-1")
+///       .getDocument(as: Player.self)
+///   print("\(p.playerID!) Health: \(p.health)")
+///
+///   // Prints: "Player: player-1, Health: 95"
+///   ````
+///
+/// - Important: Trying to encode/decode this type using encoders/decoders other than
+///   Firestore.Encoder throws an error.
+///
+/// - Important: When writing a Codable object containing an `@DocumentID` annotated field,
+///   its value is ignored. This allows you to read a document from one path and
+///   write it into another without adjusting the value here.
 @propertyWrapper
 public struct DocumentID<Value: DocumentIDWrappable & Codable>:
-  DocumentIDProtocol, Codable, StructureCodingUncodedUnkeyed {
-  var value: Value?
+  StructureCodingUncodedUnkeyed {
+  private var value: Value? = nil
 
   public init(wrappedValue value: Value?) {
+    if let value = value {
+      logWarning(for: value)
+    }
     self.value = value
   }
 
   public var wrappedValue: Value? {
     get { value }
-    set { value = newValue }
+    set {
+      if let someNewValue = newValue {
```

Comment:
So I recall being confused about introducing another `newValue` into the scope. Mainly because I didn't know if it could cause problems on  L123. Swift should be able to understand that the **non-optional** `newValue` is used in the **unwrapped scope** and the **optional** `newValue` should be set in the outer scope on L123. But I figured it was possible that an older Swift version may not know. To be extra safe, I opted for introducing a new name in the scope.

Reply:
I guess in Swift 5.7, we should even be able to use 
```
if let newValue {
}
```

Reply:
I don't think we can use this feature yet since we still need to build for devs using Xcode 13.3.1 x Swift 5.6.

Reply:
Just verified my assumption.

Reply:
Also, wrt the below https://github.com/firebase/firebase-ios-sdk/pull/10167#discussion_r983752857, the setter _should_ log if you did something like:
```swift
struct Model: Codable {
  @DocumentID var id: String = "hello"
}

let model = // ...
model.id = nil  // ⚠️ Logs a warning ⚠️
```